# HG changeset 9723 patch
# User kaf24@firebug.cl.cam.ac.uk
# Node ID 2fd3c0eb04eaeab3aa3c5d357e9b198e823b564c
# Parent  b8aa9f40d67a5f94577957b9cc9578fd8e75fb53
Subject: Support IO-port permissions bitmaps (ioperm() syscall).
From: Jan Beulich <jbeulich@novell.com>
xen-unstable changeset:   10288:28128dafbe03758434bc947002758edf60a2747e
xen-unstable date:        Wed Jun  7 13:38:08 2006 +0100

Index: head-2006-06-07/arch/x86_64/kernel/ioport-xen.c
===================================================================
--- head-2006-06-07.orig/arch/x86_64/kernel/ioport-xen.c	2006-06-09 09:32:08.000000000 +0200
+++ head-2006-06-07/arch/x86_64/kernel/ioport-xen.c	2006-06-09 09:35:05.000000000 +0200
@@ -19,6 +19,57 @@
 #include <linux/thread_info.h>
 #include <xen/interface/physdev.h>
 
+/* Set EXTENT bits starting at BASE in BITMAP to value TURN_ON. */
+static void set_bitmap(unsigned long *bitmap, unsigned int base, unsigned int extent, int new_value)
+{
+	int i;
+
+	if (new_value)
+		for (i = base; i < base + extent; i++)
+			__set_bit(i, bitmap);
+	else
+		for (i = base; i < base + extent; i++)
+			clear_bit(i, bitmap);
+}
+
+/*
+ * this changes the io permissions bitmap in the current task.
+ */
+asmlinkage long sys_ioperm(unsigned long from, unsigned long num, int turn_on)
+{
+	struct thread_struct * t = &current->thread;
+	unsigned long *bitmap;
+	physdev_op_t op;
+
+	if ((from + num <= from) || (from + num > IO_BITMAP_BITS))
+		return -EINVAL;
+	if (turn_on && !capable(CAP_SYS_RAWIO))
+		return -EPERM;
+
+	/*
+	 * If it's the first ioperm() call in this thread's lifetime, set the
+	 * IO bitmap up. ioperm() is much less timing critical than clone(),
+	 * this is why we delay this operation until now:
+	 */
+	if (!t->io_bitmap_ptr) {
+		bitmap = kmalloc(IO_BITMAP_BYTES, GFP_KERNEL);
+		if (!bitmap)
+			return -ENOMEM;
+
+		memset(bitmap, 0xff, IO_BITMAP_BYTES);
+		t->io_bitmap_ptr = bitmap;
+
+		op.cmd = PHYSDEVOP_SET_IOBITMAP;
+		op.u.set_iobitmap.bitmap   = (char *)bitmap;
+		op.u.set_iobitmap.nr_ports = IO_BITMAP_BITS;
+		HYPERVISOR_physdev_op(&op);
+	}
+
+	set_bitmap(t->io_bitmap_ptr, from, num, !turn_on);
+
+	return 0;
+}
+
 /*
  * sys_iopl has to be used when you want to access the IO ports
  * beyond the 0x3ff range: to get the full 65536 ports bitmapped
@@ -48,11 +99,3 @@ asmlinkage long sys_iopl(unsigned int ne
 
 	return 0;
 }
-
-/*
- * this changes the io permissions bitmap in the current task.
- */
-asmlinkage long sys_ioperm(unsigned long from, unsigned long num, int turn_on)
-{
-  return turn_on ? sys_iopl(3, NULL) : 0;
-}
